### PR TITLE
further refine the txn builder

### PR DESCRIPTION
**Pull request by DavidVorick**
I ditched the registration system in favor of returning an interface, as you suggested. I consider it to be a pretty big improvement.

### CODE REVIEW

**Comment by lukechampine** [comment]
`RegisterTransaction(types.Transaction{}, nil)` is so common that I would strongly consider defining a `StartTransaction` method for convenience. As far as I can tell, `RegisterTransaction` is only called with a preexisting transaction in once place, and is never called with parents at all.

File: modules/host/announce.go [hunk 1]
```diff
@@ -27,35 +27,23 @@ func ping(addr modules.NetAddress) bool {
...
-	if err != nil {
-		return err
-	}
-	txnSet, err := h.wallet.SignTransaction(id, true)
-	if err != nil {
-		return err
-	}
+	_ = txnBuilder.AddArbitraryData(append(modules.PrefixHostAnnouncement[:], announcement...))
+	txn, parents := txnBuilder.View()
+	txnSet := append(parents, txn)
```
Comment (lukechampine) [inline modules/host/announce.go]: isn't `parents` guaranteed to be nil here?

File: modules/host/negotiate.go [hunk 2]
```diff
@@ -234,7 +228,7 @@ func (h *Host) rpcContract(conn net.Conn) (err error) {
 	}
 
 	// Add the collateral to the transaction, but do not sign the transaction.
-	collateralTxn, txnID, err := h.addCollateral(unsignedTxn, terms)
+	collateralTxn, txnBuilder, err := h.addCollateral(unsignedTxn, terms)
```
Comment (lukechampine) [inline modules/host/negotiate.go]: would probably be cleaner if you created the TransactionBuilder before, and passed it `addCollateral` for modification.

File: modules/wallet.go [hunk 3]
```diff
@@ -24,115 +24,97 @@ type WalletInfo struct {
...
+	//
+	// If the whole transaction flag  is set to true, then the whole
+	// transaction flag will be set in the covered fields object. If the whole
+	// transaction flag is set to false, then the covered fields object will
+	// cover all fields that have already been added to the transaction, but
+	// will also leave room for more fields to be added.
+	Sign(wholeTransaction bool) ([]types.Transaction, error)
+
+	// View returns the incomplete transaction along with all of its parents.
+	View() (txn types.Transaction, parents []types.Transaction)
```
Comment (lukechampine) [inline modules/wallet.go]: why does `View` need to return the parents? I'm not totally clear on this, but it doesn't seem like the TransactionBuilder ever modifies the set of parents.

File: modules/host/announce.go (hunk 1 above)
Comment (DavidVorick) [inline modules/host/announce.go]: I think it makes sense to use the builder idiomatically.

File: modules/host/negotiate.go (hunk 2 above)
Comment (DavidVorick) [inline modules/host/negotiate.go]: I agree, but the host and renter files are going to be changing dramatically enough that I'm not really worried about maintaining them.

**Comment by DavidVorick** [comment]
updated

**Pull request merged**